### PR TITLE
fix(codewhisperer): support code generation for IaC languages without external plugin extensions

### PR DESCRIPTION
**Pull request by laileni-aws**
## Problem
CodeWhisperer needs terraform plugin extension to generate code suggestions for `tf` and `hcl` files, as `tf` and `hcl` files are recognized as `plaintext` by VSCode. 
## Solution

- This issue can be resolved by categorizing the files based on file format/ file extensions instead of depending on editor document language.
- Code generation will be independent of external plugin extensions.

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/codewhisperer/util/runtimeLanguageContext.ts [hunk 1]
```diff
@@ -161,7 +161,22 @@ export class RuntimeLanguageContext {
      * @param languageId : vscode language id or codewhisperer language name
```
Comment (ctlai95) [inline src/codewhisperer/util/runtimeLanguageContext.ts:161]: please update this jsdoc with the new parameter

File: src/codewhisperer/commands/onAcceptance.ts [hunk 2]
```diff
@@ -24,7 +24,12 @@ export async function onAcceptance(acceptanceEntry: OnRecommendationAcceptanceEn
      * Format document
      */
     if (acceptanceEntry.editor) {
-        const languageContext = runtimeLanguageContext.getLanguageContext(acceptanceEntry.editor.document.languageId)
+        const languageContext = runtimeLanguageContext.getLanguageContext(
+            acceptanceEntry.editor.document.languageId,
+            acceptanceEntry.editor.document.fileName.substring(
+                acceptanceEntry.editor.document.fileName.lastIndexOf('.') + 1
+            )
```
Comment (ctlai95) [inline src/codewhisperer/commands/onAcceptance.ts:31]: Looks like this is repeated in multiple places, we should probably add a utility function, `getFileExtension()` if we don't have one already

File: src/codewhisperer/util/runtimeLanguageContext.ts [hunk 3]
```diff
@@ -161,7 +161,22 @@ export class RuntimeLanguageContext {
...
      * @returns An object with a field language: CodewhispererLanguage, if no corresponding CodewhispererLanguage ID, plaintext is returned
      */
-    public getLanguageContext(languageId?: string): { language: CodewhispererLanguage } {
+    public getLanguageContext(languageId?: string, filename?: string): { language: CodewhispererLanguage } {
```
Comment (ctlai95) [inline src/codewhisperer/util/runtimeLanguageContext.ts:164]: I think instead of `filename` it should be `extension`?

File: src/test/codewhisperer/util/runtimeLanguageContext.test.ts [hunk 4]
```diff
@@ -281,8 +281,14 @@ describe('runtimeLanguageContext', function () {
...
             ['sql', 'sql'],
+            ['sql', 'sql'],
+            ['tf', 'tf'],
+            ['hcl', 'tf'],
+            ['json', 'json'],
+            ['yaml', 'yaml'],
+            ['yml', 'yaml'],
             ['plaintext', 'plaintext'],
-            ['arbitrary string', 'arbitrary string'],
+            // ['arbitrary string', 'arbitrary string'],
```
Comment (ctlai95) [inline src/test/codewhisperer/util/runtimeLanguageContext.test.ts:290]: Can this be removed?

File: src/codewhisperer/util/editorContext.ts [hunk 5]
```diff
@@ -185,7 +185,12 @@ export function validateRequest(
         req.fileContext.programmingLanguage.languageName === undefined ||
         req.fileContext.programmingLanguage.languageName.length < 1 ||
         req.fileContext.programmingLanguage.languageName.length > 128 ||
-        !runtimeLanguageContext.isLanguageSupported(req.fileContext.programmingLanguage.languageName)
+        !(
+            runtimeLanguageContext.isLanguageSupported(req.fileContext.programmingLanguage.languageName) ||
+            runtimeLanguageContext.isFileFormatSupported(
+                req.fileContext.filename.substring(req.fileContext.filename.lastIndexOf('.') + 1)
+            )
+        )
```
Comment (ctlai95) [inline src/codewhisperer/util/editorContext.ts:193]: Can we split this up? it's a bit hard to read with the nested `!(...)`

File: src/codewhisperer/util/runtimeLanguageContext.ts [hunk 6]
```diff
@@ -161,7 +161,22 @@ export class RuntimeLanguageContext {
...
+        const extensionToLanguageMap: Record<string, CodewhispererLanguage> = {
+            tf: 'tf',
+            hcl: 'tf',
+            json: 'json',
+            yaml: 'yaml',
+            yml: 'yaml',
+            // Add more mappings if needed
```
Comment (ctlai95) [inline src/codewhisperer/util/runtimeLanguageContext.ts:171]: can we add tsx/jsx also?

File: src/codewhisperer/util/runtimeLanguageContext.ts [hunk 7]
```diff
@@ -193,12 +211,21 @@ export class RuntimeLanguageContext {
...
+     * @returns true if the language is supported by CodeWhisperer otherwise false
      */
     public isLanguageSupported(languageId: string): boolean {
         const lang = this.normalizeLanguage(languageId)
         return lang !== undefined && this.normalizeLanguage(languageId) !== 'plaintext'
     }
+    /**
+     *
+     * @param fileFormat : vscode editor filecontext filename extension
+     * @returns  if the fileformat is supported by CodeWhisperer otherwise false
```
Comment (ctlai95) [inline src/codewhisperer/util/runtimeLanguageContext.ts:223]:
```suggestion
     * @returns true if the fileformat is supported by CodeWhisperer otherwise false
```

File: src/codewhisperer/util/runtimeLanguageContext.ts (hunk 6 above)
Comment (laileni-aws) [reply src/codewhisperer/util/runtimeLanguageContext.ts:171]: We can add `tsx`/`jsx` but the editor is not considering them as `plaintext`, If I am not wrong so if thats the case we do not see any difference even if we add them to this.

If editor recognizes `tsx` and `jsx` as `plaintext` then we can add them as `jsx`to`javascript` and `tsx` to `typescript`

File: src/codewhisperer/util/runtimeLanguageContext.ts (hunk 7 above)
Comment (laileni-aws) [reply src/codewhisperer/util/runtimeLanguageContext.ts:223]: Updated !

File: src/test/codewhisperer/util/runtimeLanguageContext.test.ts (hunk 4 above)
Comment (laileni-aws) [reply src/test/codewhisperer/util/runtimeLanguageContext.test.ts:290]: I need @Will-ShaoHua input on this, Can this be removed from the test case as we are testing for all the other use cases.

File: src/codewhisperer/util/editorContext.ts (hunk 5 above)
Comment (laileni-aws) [reply src/codewhisperer/util/editorContext.ts:193]: Yeah sure, I will apply` ! `negation to inside conditions which should clear the complexity.
```
    const isLanguageNameValid = (
        req.fileContext.programmingLanguage.languageName !== undefined &&
        req.fileContext.programmingLanguage.languageName.length >= 1 &&
        req.fileContext.programmingLanguage.languageName.length <= 128 &&
        (
            runtimeLanguageContext.isLanguageSupported(req.fileContext.programmingLanguage.languageName) ||
            runtimeLanguageContext.isFileFormatSupported(
                req.fileContext.filename.substring(req.fileContext.filename.lastIndexOf('.') + 1)
            )
        )
    )
```

Finally applying some logical ops of maths in coding..!

File: src/codewhisperer/util/runtimeLanguageContext.ts (hunk 1 above)
Comment (laileni-aws) [reply src/codewhisperer/util/runtimeLanguageContext.ts]: Updated!

File: src/codewhisperer/commands/onAcceptance.ts (hunk 2 above)
Comment (laileni-aws) [reply src/codewhisperer/commands/onAcceptance.ts]: Updated!

**Review by ashishrp-aws — APPROVED** [review]

File: .changes/next-release/Bug Fix-ec18555e-d716-40b5-b8d1-54dad851fdc2.json [hunk 8]
```diff
@@ -0,0 +1,4 @@
+{
+	"type": "Bug Fix",
+	"description": "CodeWhisperer supports code generation for Iaac languages without external extensions"
```
Comment (justinmk3) [inline .changes/next-release/Bug Fix-ec18555e-d716-40b5-b8d1-54dad851fdc2.json:3]: should this be IaC? what does Iaac stand for?
```suggestion
	"description": "CodeWhisperer supports code generation for IaC languages without external extensions"
```

File: src/codewhisperer/commands/onInlineAcceptance.ts [hunk 9]
```diff
@@ -73,7 +74,10 @@ export async function onInlineAcceptance(
 
     if (acceptanceEntry.editor) {
         await sleep(CodeWhispererConstants.vsCodeCursorUpdateDelay)
-        const languageContext = runtimeLanguageContext.getLanguageContext(acceptanceEntry.editor.document.languageId)
+        const languageContext = runtimeLanguageContext.getLanguageContext(
+            acceptanceEntry.editor.document.languageId,
+            getFileExtension(acceptanceEntry.editor.document)
+        )
```
Comment (justinmk3) [inline src/codewhisperer/commands/onInlineAcceptance.ts:80]: Looks like all of these calls could benefit if `getLanguageContext` accepted a document (could be an optional overload to support both ways)
```suggestion
        const languageContext = runtimeLanguageContext.getLanguageContext(editor.document)
```

File: src/codewhisperer/service/recommendationHandler.ts [hunk 10]
```diff
@@ -399,7 +403,10 @@ export class RecommendationHandler {
                     session.requestIdList,
                     sessionId,
                     page,
-                    editor.document.languageId,
+                    runtimeLanguageContext.getLanguageContext(
+                        editor.document.languageId,
+                        editor.document.fileName.substring(editor.document.fileName.lastIndexOf('.') + 1)
```
Comment (justinmk3) [inline src/codewhisperer/service/recommendationHandler.ts:408]: why not use `path.extname()` ? it is more readable, shorter, and potentially more robust.

Comment (justinmk3) [reply src/codewhisperer/service/recommendationHandler.ts:408]:
```suggestion
                        path.extname(editor.document.fileName)
```

File: src/codewhisperer/util/editorContext.ts [hunk 11]
```diff
@@ -76,6 +76,10 @@ export function getFileName(editor: vscode.TextEditor): string {
     return fileName.substring(0, CodeWhispererConstants.filenameCharsLimit)
 }
 
+export function getFileExtension(document: vscode.TextDocument): string {
+    return document.fileName.substring(document.fileName.lastIndexOf('.') + 1)
```
Comment (justinmk3) [inline src/codewhisperer/util/editorContext.ts:80]: why is this needed instead of `path.extname` ?

File: src/codewhisperer/util/runtimeLanguageContext.ts [hunk 12]
```diff
@@ -193,12 +211,21 @@ export class RuntimeLanguageContext {
...
         return lang !== undefined && this.normalizeLanguage(languageId) !== 'plaintext'
     }
+    /**
+     *
+     * @param fileFormat : vscode editor filecontext filename extension
+     * @returns  true if the fileformat is supported by CodeWhisperer otherwise false
+     */
+    public isFileFormatSupported(fileFormat: string): boolean {
+        const fileformat = this.supportedLanguageExtensionMap.get(fileFormat)
+        return fileformat !== undefined && this.supportedLanguageExtensionMap.get(fileformat) !== 'txt'
```
Comment (justinmk3) [inline src/codewhisperer/util/runtimeLanguageContext.ts:227]: `this.supportedLanguageExtensionMap.get(fileFormat)` is doing a lookup on the thing it just looked up. is that intentional?

File: src/codewhisperer/service/recommendationHandler.ts (hunk 10 above)
Comment (Will-ShaoHua) [reply src/codewhisperer/service/recommendationHandler.ts:408]: +1

File: src/codewhisperer/util/editorContext.ts [hunk 13]
```diff
@@ -181,12 +185,14 @@ export async function buildGenerateRecommendationRequest(editor: vscode.TextEdit
...
-        req.fileContext.programmingLanguage.languageName.length > 128 ||
-        !runtimeLanguageContext.isLanguageSupported(req.fileContext.programmingLanguage.languageName)
-    )
+    const isLanguageNameValid =
+        req.fileContext.programmingLanguage.languageName !== undefined &&
+        req.fileContext.programmingLanguage.languageName.length >= 1 &&
+        req.fileContext.programmingLanguage.languageName.length <= 128 &&
+        (runtimeLanguageContext.isLanguageSupported(req.fileContext.programmingLanguage.languageName) ||
+            runtimeLanguageContext.isFileFormatSupported(
+                req.fileContext.filename.substring(req.fileContext.filename.lastIndexOf('.') + 1)
```
Comment (Will-ShaoHua) [inline src/codewhisperer/util/editorContext.ts:194]: I think moving all of these logics into single one method `isLanguageSupported()` is easier to use.

File: src/codewhisperer/util/runtimeLanguageContext.ts (hunk 6 above)
Comment (Will-ShaoHua) [reply src/codewhisperer/util/runtimeLanguageContext.ts:171]: as Justin suggested, think making the method signature to be `getLanguageContext(document: vscode.TextDocument)` would be nicer and maybe rename it to be `getLanguageForDocument()`

Comment (Will-ShaoHua) [reply src/codewhisperer/util/runtimeLanguageContext.ts:171]: actually, can we also make this method returning `CodewhispererLanguage` instead of `{ language: CodewhispererLanguage }`, it's our tech debt and meaningless.

File: src/codewhisperer/util/runtimeLanguageContext.ts [hunk 14]
```diff
@@ -158,10 +158,25 @@ export class RuntimeLanguageContext {
...
+            yaml: 'yaml',
+            yml: 'yaml',
+            // Add more mappings if needed
+        }
+
+        if (languageId === 'plaintext' && fileExtension !== undefined) {
+            const languages = extensionToLanguageMap[fileExtension]
+            if (languages) {
+                return { language: languages }
+            }
```
Comment (Will-ShaoHua) [inline src/codewhisperer/util/runtimeLanguageContext.ts:178]:
```
public getLanguageContext(doc: vscode.TextDocument): CodewhispererLanguage {
  return this.normalizeLanguage(doc.languageId) ?? extensionToLanguageMap[fileExt] ?? 'plaintext'
}
```

File: src/test/codewhisperer/util/runtimeLanguageContext.test.ts (hunk 4 above)
Comment (Will-ShaoHua) [reply src/test/codewhisperer/util/runtimeLanguageContext.test.ts:290]: any reason why we need to comment this out?

Comment (Will-ShaoHua) [reply src/test/codewhisperer/util/runtimeLanguageContext.test.ts:290]: if it's not causing compile error or any error I don't get the reason why we have this change here

File: src/codewhisperer/util/telemetryHelper.ts [hunk 15]
```diff
@@ -99,10 +99,9 @@ export class TelemetryHelper {
         requestIdList: string[],
         sessionId: string,
         paginationIndex: number,
-        languageId: string,
+        languageContext: CodewhispererLanguage,
```
Comment (Will-ShaoHua) [inline src/codewhisperer/util/telemetryHelper.ts:102]: nit: `language: CodewhispererLanguage`

File: src/test/codewhisperer/util/runtimeLanguageContext.test.ts (hunk 4 above)
Comment (laileni-aws) [reply src/test/codewhisperer/util/runtimeLanguageContext.test.ts:290]: Can I know what it does ?

File: src/codewhisperer/util/runtimeLanguageContext.ts (hunk 12 above)
Comment (Will-ShaoHua) [reply src/codewhisperer/util/runtimeLanguageContext.ts:227]: prefer having single one interface `isLanguageSupported` but we can make the signature to be `isLanguageSupported(documetn: vscode.TextDocument)`

File: src/test/codewhisperer/util/runtimeLanguageContext.test.ts (hunk 4 above)
Comment (Will-ShaoHua) [reply src/test/codewhisperer/util/runtimeLanguageContext.test.ts:290]: it's just a test case. 
```
public mapToRuntimeLanguage<
        T extends codewhispererClient.ListRecommendationsRequest | codewhispererClient.GenerateRecommendationsRequest
    >(request: T): T
```

type of `T.fileContext.programmingLanguage.languageName` is `string`, we can't gurantee it's something we must know in advance

Comment (Will-ShaoHua) [reply src/test/codewhisperer/util/runtimeLanguageContext.test.ts:290]: of course you can remove it IF we update and make our logic more robust by passing type of `CodewhispererLanguage` or something which will show error at compile time.

Comment (Will-ShaoHua) [reply src/test/codewhisperer/util/runtimeLanguageContext.test.ts:290]: although this method `mapToRuntimeLanguage` is another method we should refactor and remove, it's not covered by this PR so let's keep it as is now

File: src/codewhisperer/util/editorContext.ts (hunk 13 above)
Comment (laileni-aws) [reply src/codewhisperer/util/editorContext.ts:194]: Yes, but this is taking a cascade effect of changing 6 files, If Yes, I can change in those files too..

File: src/codewhisperer/util/runtimeLanguageContext.ts (hunk 12 above)
Comment (laileni-aws) [reply src/codewhisperer/util/runtimeLanguageContext.ts:227]: > this.supportedLanguageExtensionMap.get(fileFormat) is doing a lookup on the thing it just looked up. is that intentional?

I think we did not check for file extension before 

```
    (runtimeLanguageContext.isLanguageSupported(req.fileContext.programmingLanguage.languageName) ||
             runtimeLanguageContext.isFileFormatSupported(
                 req.fileContext.filename.substring(req.fileContext.filename.lastIndexOf('.') + 1)
             ))
```

In this code we check for language and then we check for file extension if anything matches this is treated as a validLanguage name

File: src/codewhisperer/util/runtimeLanguageContext.ts (hunk 14 above)
Comment (laileni-aws) [reply src/codewhisperer/util/runtimeLanguageContext.ts:178]: We can not use `(doc: vscode.TextDocument)` as parameter because this `getLanguageContext` is used in different parts of code, example in this we pass `request: codewhispererClient.ListRecommendationsRequest | codewhispererClient.GenerateRecommendationsRequest` so we can not generalize into document as parameter
```      
                runtimeLanguageContext.getLanguageContext(
                    request.fileContext.programmingLanguage.languageName,
                    request.fileContext.filename.substring(request.fileContext.filename.lastIndexOf('.') + 1)
                ).language
```

File: src/codewhisperer/util/editorContext.ts [hunk 16]
```diff
@@ -76,6 +76,10 @@ export function getFileName(editor: vscode.TextEditor): string {
     return fileName.substring(0, CodeWhispererConstants.filenameCharsLimit)
 }
 
+export function getFileExtension(document: vscode.TextDocument): string {
```
Comment (justinmk3) [inline src/codewhisperer/util/editorContext.ts:79]: this function is not needed.

File: src/codewhisperer/util/runtimeLanguageContext.ts (hunk 12 above)
Comment (Will-ShaoHua) [reply src/codewhisperer/util/runtimeLanguageContext.ts:227]: correct, we didn't check file ext before, I mean can we add these layer of logic into existing `isLanguageSupport()` method instead exposing a new API? 

```
public isLanguageSupport(document: vscode.TextDocument) {
    const Lang = this.normalizedLanguage(document.languageId)
    if (this.supportedLanguageMap.get(normalizedLang)) {
        return true
    } else {
         // check file ext
    }
}
```

Comment (Will-ShaoHua) [reply src/codewhisperer/util/runtimeLanguageContext.ts:227]: the reasoning is that basically we will always need to 
1. check language id if present
2. if step 1 fails,  then go check file ext.

in this case we don't really split these 2 into 2 API and callers have to explicitly call these 2 APIs. 1 is enough and should be more error prone IMO.

File: src/codewhisperer/util/runtimeLanguageContext.ts (hunk 14 above)
Comment (Will-ShaoHua) [reply src/codewhisperer/util/runtimeLanguageContext.ts:178]: I see.

File: src/codewhisperer/util/editorContext.ts (hunk 13 above)
Comment (Will-ShaoHua) [reply src/codewhisperer/util/editorContext.ts:194]: ok make sense, it's something should be fixed with a refactor. In that case it's fine here.

**Comment by Will-ShaoHua** [comment]
LGTM. My comments should mostly be take care in CW refactor PRs instead of this one.

**Review by Will-ShaoHua — APPROVED** [review]

File: src/codewhisperer/util/runtimeLanguageContext.ts (hunk 12 above)
Comment (laileni-aws) [reply src/codewhisperer/util/runtimeLanguageContext.ts:227]: Yes we can add this to `isLanguageSupport` but as I mentioned below we need more refactoring of the code in different files. Let me take up and do this in this PR itself

File: src/codewhisperer/commands/onInlineAcceptance.ts (hunk 9 above)
Comment (laileni-aws) [reply src/codewhisperer/commands/onInlineAcceptance.ts:80]: responded below.

Comment (justinmk3) [reply src/codewhisperer/commands/onInlineAcceptance.ts]: Can't that be solved with an overload?

**Review by justinmk3 — APPROVED** [review]

File: src/codewhisperer/commands/onInlineAcceptance.ts (hunk 9 above)
Comment (laileni-aws) [reply src/codewhisperer/commands/onInlineAcceptance.ts]: `getLanguageContext` needs more refactoring along with other functions `mapToRuntimeLanguage` to do this.
